### ISSUE TITLE
Reproducing the Results of Table 1

### ISSUE THREAD

**Issue by Vujas-Eteph** [hkchengrex/XMem#73]
Hi, Very nice work!
I like your previous projects and papers (e.g., MiVOS, STCN), and I currently use them in my project!

Regarding the XMem paper, I have a few questions about **Table 1** (Quantitative comparisons on the Long-time Video dataset) that I would like to reproduce. However, I am not sure I understood the configurations you used (other than limiting the memory to 50 "memory slots" as in the AFB-URR paper) for the space-time memory network models (i.e., STM, MiVOS, AOT, STCN).
I would be very grateful if you could elaborate on this :)

Specifically, I would like to know the following:
- (1) Since you restrict the number of frames stored in the external memory, did you follow a first-in-first-out strategy? Did you keep the annotated image in memory when segmenting the entire sequence? Did you use the previous adjacent frame?
- (2) I understand you limited the insertion frequency to make these methods suitable for long-term scenarios; what value did you use here? Is it also 10, as in XMem? Or did you use a higher value (e.g., 50)?
- (3) If I read correctly, only the MiVOS implementation [`s012`](https://github.com/hkchengrex/Mask-Propagation) performs better when trained with BL30K, while STCN performs better when using the [weights](https://github.com/hkchengrex/STCN) of `s03` instead of `s012`?
- (4) Are there any other changes not mentioned in the paper that are crucial to reproduce the results from Table 1?

I would appreciate any feedback!
With kind regards!

**Comment by hkchengrex** [comment]
Hi, thanks a lot for your interest.

From the AFB-URR paper: "Because the total frames that STM can store is fixed, when the video length becomes longer, STM has to increase the key frame interval and has a higher chance of missing important frames and information..."

Answer to (1) and (2): we decrease the memory frequency such that we use exactly 50 memory frames for STM (applied per-video). We apply the same limit to STCN/MiVOS. For methods that use a significantly different amount of memory over time (Fig. 1), we observe the memory usage for 50-frame STM and try to convert that to the best equivalent memory frame limit and relax the memory frequency in the same way.

(3) IIRC both STCN and XMem have a performance drop when BL30K is used. I did not try MiVOS model without BL30K pretraining.

(4) Implementation-wise, STM/MiVOS/STCN loads all frames to the CPU before processing which can be problematic. Some refactoring on data loading is required but that does not change the core behavior of the algorithm. If you have enough CPU memory, it should be reproduced just by taking the code of STM and decreasing its memory frequency.

**Comment by Vujas-Eteph** [comment]
Thank you for the quick answer!

Last question before closing this issue. Could you elaborate on how you measured the memory usage/increase? (I suppose that this part of the code is a key element in performing this evaluation?)
https://github.com/hkchengrex/XMem/blob/5fe108138d457774f91ace362ecf97b9491f81d9/eval.py#L246-L249
Did you use a particular sequence (e.g., blueboy, dressage, rat) of the Long Video Dataset for this part?

Best!

**Comment by hkchengrex** [comment]
It is mentioned in Section 4.1.
We used blueboy.

**Comment by Vujas-Eteph** [comment]
Alright, I'll retake a look at that part.
Thanks!
